File: src/vca/state.py
"""Agent 状态定义 - LangGraph 的状态管理"""

from __future__ import annotations

from typing import Annotated, Any, TypedDict
from operator import add
from langchain_core.messages import BaseMessage


class PlanStep(TypedDict):
    """计划中的一个步骤"""
    description: str
    tool_name: str | None
    tool_args: dict[str, Any] | None
    status: str  # pending | running | done | failed


class AgentState(TypedDict):
    """
    LangGraph Agent 的全局状态。

    通过 TypedDict + Annotated 定义 reducer:
    - messages 使用 add 合并，实现消息累积
    - 其他字段直接覆盖 (默认 reducer)
    """
    # 消息历史 (自动累积)
    messages: Annotated[list[BaseMessage], add]

    # 当前执行计划
    plan: list[PlanStep]

    # 当前步骤索引
    current_step: int

    # 工具执行结果
    tool_results: list[dict[str, Any]]

    # 迭代计数器 (防止无限循环)
    iteration: int

    # 最终响应
    final_response: str

    # 工作空间路径
    workspace_dir: str

    # --- AskUser 挂起字段 ---
    # 当 LLM 调用 ask_user 时，挂起问题详情
    # 格式: {"header": "...", "question": "...", "options": "...", "is_multi": bool, "tool_call_id": "..."}
    # None 表示无挂起问题
    pending_question: dict[str, Any] | None

    # --- 上下文裁剪摘要 ---
    # 当对话超长被裁剪时，LLM 生成的语义摘要（累积）
    summary_history: str


def create_initial_state(workspace_dir: str) -> AgentState:
    """创建初始状态"""
    return AgentState(
        messages=[],
        plan=[],
        current_step=0,
        tool_results=[],
        iteration=0,
        final_response="",
        workspace_dir=workspace_dir,
        pending_question=None,
        summary_history="",
    )

File: src/agent/state.ts
/**
 * Agent 状态定义 - LangGraph 的状态管理
 */
import { Annotation } from "@langchain/langgraph";
import type { BaseMessage } from "@langchain/core/messages";

export interface PlanStep {
  description: string;
  status: string; // pending | in_progress | done | failed
}

export interface PendingQuestion {
  header: string;
  question: string;
  options: string[];
  is_multi: boolean;
  tool_call_id: string;
}

export interface LlmUsage {
  input_tokens: number;
  output_tokens: number;
  total_tokens: number;
  duration_ms: number;
}

export interface ToolUsage {
  count: number;
  duration_ms: number;
}

export interface AgentState {
  messages: BaseMessage[];
  plan: PlanStep[];
  current_step: number;
  tool_results: Record<string, unknown>[];
  iteration: number;
  final_response: string;
  workspace_dir: string;
  pending_question: PendingQuestion | null;
  summary_history: string;
  llm_usage: LlmUsage | null;
  tool_usage: ToolUsage | null;
  session_id?: string;
}

/**
 * LangGraph 状态注解。
 * - messages 使用 concat reducer 累积
 * - 其他字段默认覆盖
 */
const overwrite = <T>(defaultValue: () => T) => ({
  reducer: (_current: T, updated: T) => updated,
  default: defaultValue,
});

export const StateAnnotation = Annotation.Root({
  messages: Annotation<BaseMessage[]>({
    reducer: (left, right) => left.concat(right),
    default: () => [],
  }),
  plan: Annotation<PlanStep[]>(overwrite<PlanStep[]>(() => [])),
  current_step: Annotation<number>(overwrite<number>(() => 0)),
  tool_results: Annotation<Record<string, unknown>[]>(overwrite<Record<string, unknown>[]>(() => [])),
  iteration: Annotation<number>(overwrite<number>(() => 0)),
  final_response: Annotation<string>(overwrite<string>(() => "")),
  workspace_dir: Annotation<string>(overwrite<string>(() => "")),
  pending_question: Annotation<PendingQuestion | null>(overwrite<PendingQuestion | null>(() => null)),
  summary_history: Annotation<string>(overwrite<string>(() => "")),
  llm_usage: Annotation<LlmUsage | null>(overwrite<LlmUsage | null>(() => null)),
  tool_usage: Annotation<ToolUsage | null>(overwrite<ToolUsage | null>(() => null)),
});

export function createInitialState(workspaceDir: string): AgentState {
  return {
    messages: [],
    plan: [],
    current_step: 0,
    tool_results: [],
    iteration: 0,
    final_response: "",
    workspace_dir: workspaceDir,
    pending_question: null,
    summary_history: "",
    llm_usage: null,
    tool_usage: null,
    session_id: undefined,
  };
}
